tests/components/balboa/test_climate.py

```diff
@@ -0,0 +1,264 @@
+"""Tests of the climate entity of the balboa integration."""
+
+from unittest.mock import patch
+
+from homeassistant.components.balboa.const import DOMAIN as BALBOA_DOMAIN
+from homeassistant.components.climate.const import (
+    ATTR_FAN_MODE,
+    ATTR_HVAC_ACTION,
+    ATTR_HVAC_MODES,
+    ATTR_MAX_TEMP,
+    ATTR_MIN_TEMP,
+    ATTR_PRESET_MODE,
+    ATTR_PRESET_MODES,
+    CURRENT_HVAC_HEAT,
+    CURRENT_HVAC_IDLE,
+    FAN_HIGH,
+    FAN_LOW,
+    FAN_MEDIUM,
+    FAN_OFF,
+    HVAC_MODE_AUTO,
+    HVAC_MODE_HEAT,
+    HVAC_MODE_OFF,
+    SUPPORT_FAN_MODE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+)
+from homeassistant.const import ATTR_TEMPERATURE, TEMP_FAHRENHEIT
+from homeassistant.core import HomeAssistant
+from homeassistant.setup import async_setup_component
+
+from . import init_integration_mocked
+
+from tests.components.climate import common
+
+FAN_SETTINGS = [
+    FAN_OFF,
+    FAN_LOW,
+    FAN_MEDIUM,
+    FAN_HIGH,
+]
+
+HVAC_SETTINGS = [
+    HVAC_MODE_HEAT,
+    HVAC_MODE_OFF,
+    HVAC_MODE_AUTO,
+]
+
+ENTITY_CLIMATE = "climate.fakespa_climate"
+
+
+async def test_spa_defaults(hass: HomeAssistant):
+    """Test supported features flags."""
+
+    config_entry = await init_integration_mocked(hass)
+
+    await async_setup_component(hass, BALBOA_DOMAIN, config_entry)
+    await hass.async_block_till_done()
+
+    state = hass.states.get(ENTITY_CLIMATE)
+
+    assert (
+        state.attributes["supported_features"]
+        == SUPPORT_TARGET_TEMPERATURE | SUPPORT_PRESET_MODE
+    )
+    assert state.state == HVAC_MODE_HEAT
+    assert state.attributes[ATTR_MIN_TEMP] == 10.0
+    assert state.attributes[ATTR_MAX_TEMP] == 26.5
+    assert state.attributes[ATTR_PRESET_MODE] == "Ready"
+    assert state.attributes[ATTR_HVAC_ACTION] == CURRENT_HVAC_IDLE
+
+
+async def test_spa_with_blower(hass: HomeAssistant):
+    """Test supported features flags."""
+
+    config_entry = await init_integration_mocked(hass)
+
+    await async_setup_component(hass, BALBOA_DOMAIN, config_entry)
+    await hass.async_block_till_done()
+
+    spa = hass.data[BALBOA_DOMAIN][config_entry.entry_id]
```

Comment:
We don't want to access the internals (`hass.data`) of the integration in tests .   Patch the library and access the mocked version instead.

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

Reply:
how am I supposed to get the client then, or are you saying I can't access that at all in the tests?

Reply:
patch("homeassistant.components.balboa.BalboaSpaWifi") and replace it with your mock

Reply:
To avoid having to repeat the pattern in multiple tests, you can do something like:


https://github.com/home-assistant/core/blob/dev/tests/components/flux_led/__init__.py#L129
https://github.com/home-assistant/core/blob/dev/tests/components/bond/common.py#L184

Reply:
OK, but if I do that, I'll have to re-implement basically the entire library in the mock,  Then what exactly am I testing anymore?  If I'm not actually testing how it interacts with the library, then am I just making sure the bits flip the way i tell them to when I tell them to in ha?

Reply:
The goal is to test the integration code in HA not the library itself.

When possible, we shouldn't be testing that the library makes the expected calls to the device in HA.  Those tests should live in the library codebase.

Reply:
OK, I'll have to re-look at this entirely.  I thought the whole point of the tests was to make sure the integration still worked with the library.  I'm not trying to check that the library makes calls to the device, because I've mocked the connections all out, just that HA sees the proper values after the library hands them back.

If that's not the intent of the tests, then my whole test_climate.py is non-helpful.

Reply:
The tests seem fine.   You could avoid accessing `hass.data[BALBOA_DOMAIN][config_entry.entry_id]` with by using the `patch` to only capture the object instead of replacing it since you are already creating a partial mock of the PyPI code with `init_integration_mocked`

Reply:
but then I'm not interacting with the same client that hass is interacting with?  I'm actually completely lost here as to what you are saying to do.

Somehow, I have to do:
spa.settemp = 15.4
and then assert state.attributes.get(ATTR_TEMPERATURE) == 15.0

If the two things aren't accessing the same object, then how is that even possible?

Or like I said, I must be completely misunderstanding you, which is more likely.

Reply:
ok, after a few hours I finally get this, I think...  sorry for being dense here.

Reply:
No worries. Feel free to request review once you change the first one so we can make sure we are on the same page before doing the others